### PR TITLE
KYLIN-3983 Add extra metadata for measure. Add test case and example/…

### PR DESCRIPTION
**Pull request by Qsimple**
…xample/test_case_data for MeasureManager

UT pass
IT pass

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: core-cube/src/main/java/org/apache/kylin/measure/CubeL2Cache.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+
+    public CubeL2Cache(CubeManager cubeMgr) {
+        this.cubeMgr = cubeMgr;
+    }
+
+    // cat
+    public List<MeasureInstance> getMeasuresOnCubeAllowMiss(String cubeName) {
+        CubeCache cache = cubeCaches.get(cubeName);
+        cache = cache == null ? reloadCacheAllowMiss(cubeName) : cache;
```
Comment (nichunen) [inline core-cube/src/main/java/org/apache/kylin/measure/CubeL2Cache.java]: redundant code

File: core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+
+public class MeasureManager {
+
+    public static final Serializer<MeasureInstance> MEASURE_SERIALIZER = new JsonSerializer<>(MeasureInstance.class);
+
+    private static final Logger LOG = LoggerFactory.getLogger(MeasureManager.class);
+    public static final String MEASURE = "measure";
+    public static final String CUBE = "cube";
+    public static final String CUBE_DESC = "cube_desc";
+    public static final String CUBE_MEASURE = "cube_measure";
```
Comment (nichunen) [inline core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: what is "cube_measure" for?

File: core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+            String cubeName = r.getRealization();
+            initialCubeMeasure(cubeName);
+        }
+    }
+
+    private static void initialCubeMeasure(String cubeName) throws IOException {
+        KylinConfig config = KylinConfig.getInstanceFromEnv();
+        CubeInstance cube = CubeManager.getInstance(config).getCube(cubeName);
+        if (null == cube) {
+            System.out.println(String.format(Locale.ROOT, "Can't find cube by project: %s, cube: %s.", cubeName));
```
Comment (nichunen) [inline core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: why not logged?

**Comment by Qsimple** [comment]
Yes,  I agree with you.  Another PR will be prepared to replace this one.  Tks for you,  chunen

File: core-cube/src/main/java/org/apache/kylin/cube/CubeManager.java [hunk 4]
```diff
@@ -238,7 +239,7 @@ public CubeInstance createCube(String cubeName, String projectName, CubeDesc des
 
             ProjectManager.getInstance(config).moveRealizationToProject(RealizationType.CUBE, cubeName, projectName,
                     owner);
-
+            getMeasureManager().createMeasuresOnCube(cube);
```
[the remark this answers is not in the document]
Comment (Qsimple) [reply core-cube/src/main/java/org/apache/kylin/cube/CubeManager.java]: I don't understand this.  why need to put "Create A" into "Update B".

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 5]
```diff
@@ -2184,4 +2184,32 @@ public boolean isStreamingStandAloneMode() {
     public String getLocalStorageImpl() {
         return getOptional("kylin.stream.settled.storage", null);
     }
+
+    public String getHBaseColumnQualifierPrefix() {
```
[the remark this answers is not in the document]
Comment (Qsimple) [reply core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java]: ok

File: core-cube/src/main/java/org/apache/kylin/measure/CubeL2Cache.java (hunk 1 above)
Comment (Qsimple) [reply core-cube/src/main/java/org/apache/kylin/measure/CubeL2Cache.java]: ok

File: core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java (hunk 2 above)
Comment (Qsimple) [reply core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: When Listener receive "cube_measure", MeasureManager will batch-process the MeasureInstances on the cube

File: core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java (hunk 3 above)
Comment (Qsimple) [reply core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: ok

File: core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java (hunk 2 above)
Comment (nichunen) [reply core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: > When Listener receive "cube_measure", MeasureManager will batch-process the MeasureInstances on the cube

I found you only add "/measure" as metadata in org.apache.kylin.common.persistence.ResourceStore, but no "cube_measure"

Comment (Qsimple) [reply core-cube/src/main/java/org/apache/kylin/measure/MeasureManager.java]: Yes，cube_measure means the Measure Instances on same  cube have been updated and needed to be reloaded. The most methods in MeasureManager will batch process on  Measure instances on same cube and will publish an cube_measure entitie's events with cube's name. MeasureManager can also modify single Measure instance and trigger a 'measure' event,  which is seldomly called. I ll remove cube and cube_dese which doesn't matter any more

**Comment by Qsimple** [comment]
new pr https://github.com/apache/kylin/pull/638

**Pull request closed without merging**